### ISSUE TITLE
Built-in confirmation page

### ISSUE THREAD

**Issue by ebruchez** [orbeon/orbeon-forms#2985]
I thought we already had an RFE for this.

See also #2514, #2458.

See for example this screenshot. It combines:

- appearing as the last step of a wizard
- saying that it's a confirmation page
- the ability to download a PDF
- the ability to email a receipt

More can certainly be done.

![confirmation](https://cloud.githubusercontent.com/assets/105769/19872540/87d75e20-9f77-11e6-8e28-482e0fe15447.png)

**Comment by ebruchez** [comment]
[+1 from user](https://mail.google.com/mail/u/0/#search/kamal/FMfcgxvzMBlhwTwBJMSqzwmGMMBTPcpq)

**Comment by avernet** [comment]
[+1 from user](http://discuss.orbeon.com/Result-dialog-display-PDF-url-link-td4664433.html)

**Comment by ebruchez** [comment]
One way to do this would be with a new Form Runner action called `confirmation`. This action would navigate to the confirmation page by doing a POST of the PDF URL. So you would write:

```
...
then send
then confirmation
```

This would be a shortcut for something like:

```
...
then send
then send(method = "post", content = "pdf-url", uri = "/fr/confirmation")
```

The `confirmation` could take options:

- PDF download ("Print Application") like in above
- email button

The title and text on the page should be easily changeable/localizable.

**Comment by avernet** [comment]
Other than the buttons on (or at the bottom of) the confirmation page, the content of the confirmation page can be very much like the content of the existing confirmation email. So, it would be nice if we could use a similar mechanism and user interface to define both, and if this could be done from Form Builder.

**Comment by ebruchez** [comment]
The "page"  might also not be a separate page, SPA-style. It would just be that the form would show the confirmation content.

See also #3767 to lock the form after confirmation.

**Comment by ebruchez** [comment]
+1 from me today

**Comment by ebruchez** [comment]
If we have #3767 (ability to lock a form after submission) then, in the SPA spirit, we should just have a confirmation "view" of the page. We don't need to navigate to another page on the server.

**Comment by avernet** [comment]
[+1 from customer](https://3.basecamp.com/3600924/buckets/10984927/messages/3599044778)

**Comment by avernet** [comment]
[+1 from customer](https://3.basecamp.com/3600924/buckets/3893142/messages/7131267345)

**Comment by ebruchez** [comment]
With #7073, we now have a good extension mechanism which allows us to easily make new modes.

We had thought, for the confirmation page, that it should be able to do this:

- show a standard message
- and/or show a custom formatted text which can refer to the form's data
- have its own buttons

This can be done with a new mode:

- new `confirmation` mode
- implemented by `<fr:confirmation>`
    - consists of a rich text field which resolves control values and formulas with the usual template mechanism

This should be a low-hanging fruit once #7073 is done.

**Comment by ebruchez** [comment]
Steps:

- [x] `fr:confirmation`
    - [x] create component
    - [x] reads its template from form metadata
        - similar to email templates?
        - only 1 template for now?
    - [x] select template based on language
    - [x] PDF download generation/link
    - [ ] other attachments links?
        - not a priority
- [x] Form Builder UI
    - ability to edit confirmation page template
    - see thinking below: reuse Email Settings dialog
- [x] doc

Example:

```xml
<confirmation>
    <templates>
        <template name="default" xml:lang="en">
            <body>
                Thank you {$name}! You have {$my-foo} widgets.
            </body>
        </template>
    </templates>
    <parameters>
        <param type="ControlValueParam">
            <name>my-name</name>
            <controlName>name</controlName>
        </param>
        <param type="ExpressionParam">
            <name>my-foo</name>
            <expr>42</expr>
        </param>
    </parameters>
</confirmation>
```

Or we could scope mode-specific configuration differently, such as:

```xml
<modes>
    <fr:confirmation>
        ....
```

**Comment by ebruchez** [comment]
This would look very much like the Email Templates UI, but simpler. Ideally, we would reuse most of the existing code.

**Comment by ebruchez** [comment]
Practical issue: how do we write code which handles parameters? For LHHA and a few others, we do this at form transformation time. But in a mode implementation, we can't do this. We probably need to reuse `EmailMetadataParsing` and related.

However, we wouldn't want to parse the template definitions all the time. When this is processed ahead of time, the result is an XPath expression which can be cached. If we have a function which takes a template, in the metadata and possibly in properties, the result should be a cacheable expression. But against what could it be cached?

Alternatively, could we still, statically, access the metadata and properties? We have as a precedent `FRComponentParamSupport.fromMetadataAndProperties()`, used statically for `fr:keep-if-param-non-blank`. But how to express it in the component?

Mmh, can we call a native function at XSLT transformation time in the XBL component?

```xml
<xf:output
    mediatype="text/html"
    value="{frf:makeResourceWithParams(....)}">
```

That native function, like `fromMetadataAndProperties()`, would have the means to statically access metadata and properties.

**Comment by ebruchez** [comment]
Several features that applies to an email template apply to the confirmation page:

- subject (title)
- body (message)
- PDF attachment
- other file attachments
- ability to pick a specific "template" depending on conditions

This makes sense, since an email is itself often a confirmation email.

Excluded are:

- recipients, etc.
- headers

Could a confirmation page "simply" be configured to use an email template? If that's the case, we might want eventually to rename/merge the two, but as a first implementation, the process could be:

- create an (email) template
- specify that you want to use that template for the confirmation page

A template could be configured, in the UI, with two additional checkboxes:

- Allow use as email template
- Allow use as confirmation page template

As a start, we could omit these checkboxes. Alternatively, like we do for sections/grids, the same dialog could be used to edit email templates and confirmation templates, but would show "Email Templates" and "Confirmation Templates" in its title, and omit parts that are not relevant for one or the other.

The `fr:confirmation` mode component then only has to pick and display a template. But which? Currently, the `change-mode()` process action does not support taking parameters.  But you'd like, intuitively:

```
change-mode(template = "my-template")
```

If you already have email templates, you probably want a different template for the confirmation page. So it seems important to separate email and confirmation templates.

Possible plan:

- separate email/confirmation templates
- confirmation templates are stored separately from email templates
- modify and reuse Email Settings dialog for confirmation templates, but make it a different dialog from the user's perspective
- for now, the template name is not used and templates can only be selected by language
- the use of a confirmation page must still be done in a button/process
- global templates, configured via properties, can come later

**Comment by ebruchez** [comment]
RESOLUTION: We will reuse the Email Settings dialog.

**Comment by ebruchez** [comment]
We have at least two way to evaluate title/message (subject/body):

1. From the XSLT function
2. From the Scala code used for email templates. We have `EvaluatedParams.fromEmailMetadata()` which takes a template and parameter descriptions, and returns `EvaluatedParams`, from which we can evaluate the template.

To address the caching concerns, we could do the following in `fr:confirmation`:

- at the opportune moment, imperatively evaluate the template by calling Scala code
- store the result into a local instance
- output the result from local instances

We probably don't yet have an appropriate event which could trigger this, but we could add one. This should happen after all form load actions have run.

**Comment by ebruchez** [comment]
We'd like third-parties to also be able to do something similar, for custom confirmation pages. Is this is a satisfactory solution?

- documented `fr-controls-and-actions-ready` (name TBD) event on the component
- function to evaluate confirmation mode templates

**Comment by ebruchez** [comment]
Email code uses an `FormRunnerDocContext`, specifically an `InDocFormRunnerDocContext`, which is used for:

- getting namespaces in scope on the main `<xf:model>`
    - we could find a way to obtain that from the live document
- but also for `searchControlsTopLevelOnly()`, which relies on having the source of a form definition
    - this is a little harder to 

So we can't use this directly.

Before 2025.1, building the email content was done in a separate service, which did not directly have access to the live form, so it used the form definition source. When moving the code to Scala, this was not changed. We call `PersistenceApi.readPublishedFormDefinition()` to get the form definition source. But this would technically no longer be needed and we could instead rely on calling the live form.

`FormRunnerDocContext` does not work as an abstraction for what we are trying to do here. For example, with a live document, a `ControlValue` parameter should get the value from `fr-form-instance`, etc.

**Comment by ebruchez** [comment]
How should the PDF download and other "rendered formats" be handled? If the previous page created a PDF file, the URL would be stored as a temporary file. The URLs are stored into `fr-urls-instance`.

Ideally, the exact same file should be made available, and not a new generation of the PDF file, when landing on the confirmation page. This means that we should pass around `fr-urls-instance` during mode changes.

`fr-urls-instance` is used for:

- `send()`
- `email()`

This is not used by:

- `open-rendered-format()`

Are there scenarios where we need to produce a PDF multiple times?

- A "PDF" button on the page, in particular, will regenerate the PDF every time it is activated, and this makes sense, as you want an up to date result every time you use the button.
- But when sending/submitting/emailing a PDF in quick succession, you want to keep a consistent copy around.

If the confirmation page needs rendered formats, we need:

- passing around `fr-urls-instance` during mode changes, or at least this mode change
- a new function such as `fr:get-or-create-rendered-format-url(use-pdf-template = true(), pdf-template-name = 'bar', pdf-template-lang = "en")`

**Comment by ebruchez** [comment]
The feature now supports:

- new `confirmation` mode
- message title/body
- passing rendered formats information between modes
- ability to download the PDF specifically
- Form Builder UI to create confirmation page templates

Missing:

- ability to configure/download other attachments
- other?

**Comment by ebruchez** [comment]
[Documented](https://doc.orbeon.com/form-runner/form-runner-pages/confirmation-page)

**Comment by ebruchez** [comment]
[Documented Form Runner functions](https://doc.orbeon.com/xforms/xpath/extension-functions/extension-form-runner#rendered-format-functions)
